Add validity checks for HTTP datagrams

FlorianHockmann opened the pull request:
Added overloads for the `CalculateIsValid ` method in `HttpRequestDatagram` and `HttpResponseDatagram`. This allows to check whether a HTTP datagram is actually a valid datagram before parsing it.

According to the RFCs of HTTP/1.0 and 1.1 a HTTP request must contain at least a method, an URI, and the version. So `HttpRequestDatagram.CalculateIsValid()` returns `true` if the parsed method, URI, and version are not `null` (or empty in case of the URI string).

For responses, the RFCs require at least the version, status code, and a reason phrase. Since the reason phrase can be basically any string I excluded it from the validity check and just check the version and status code. So `HttpResponseDatagram.CalculateIsValid()` returns `true` if the parsed version and the status code are not `null`.

This pull request also includes unit tests for the different cases of missing / present values for the mentioned elements.

bricknerb commented on PcapDotNet/src/PcapDotNet.Packets.Test/HttpTests.cs:
```diff
@@ -470,6 +470,38 @@ public void HttpRequestMethodBadKnownTest()
+            Assert.IsFalse(packet.Ethernet.IpV4.Tcp.Http.IsValid);
+        }
+
         private static void TestHttpRequest(string httpString, string expectedMethodString = null, string expectedUri = null, HttpVersion expectedVersion = null, HttpHeader expectedHeader = null, string expectedBodyString = null)
```
Also add tests in RandomHttpTest

bricknerb commented on PcapDotNet/src/PcapDotNet.Packets/Http/HttpRequestDatagram.cs:
```diff
@@ -46,6 +46,18 @@ public override ILayer ExtractLayer()
+        /// <summary>
+        /// An HTTP Request is valid if it contains a method, an URI, and a version.
+        /// </summary>
+        protected override bool CalculateIsValid()
```
I suggest creating a separate property for this instead of overriding CalculateIsValid().
The name of the property can be something like "IsValidStart".
The reason is that an HttpDatagram can be the continuation of a previous HttpDatagram and so doesn't have to start with the first line to be valid.

FlorianHockmann replied:
You're right. I didn't think of that because I reassemble the TCP datagrams before parsing the application layer.

FlorianHockmann commented:
I am not sure if I got the `RandomHttpTest` right. It now checks that `IsValidStart` is true for all HttpDatagrams that contain a header. This should be the case at least for all packets generated during this test due to the way the `RandomHttpExtensions` work.
But I am of course open for any suggestions for a better test.

bricknerb commented on PcapDotNet/src/PcapDotNet.Packets/Http/HttpResponseDatagram.cs:
```diff
@@ -50,6 +50,14 @@ public override ILayer ExtractLayer()
         }
 
+        /// <summary>
+        /// A HTTP response has a valid start if it contains a version and a status code.
```
Replace "A HTTP" with "An HTTP".

bricknerb commented on PcapDotNet/src/PcapDotNet.Packets.Test/HttpTests.cs:
```diff
@@ -470,6 +471,86 @@ public void HttpRequestMethodBadKnownTest()
+            var packet = BuildPacket("HTTP/1.0  OK \r\n\r\n");
+
+            Assert.IsFalse(packet.Ethernet.IpV4.Tcp.Http.IsValidStart);
+        }
```
If you only care about IsValidStart in these tests, you could put "IsValidStart" as part of the test names.
Otherwise you could test other properties for these use cases.

FlorianHockmann commented:
Thank you for creating such a great library!

I fixed the typo and renamed the tests to follow a common schema that is more descriptive and fits in better with the other tests.
